### ISSUE TITLE
Accession

### ISSUE THREAD

**Issue by rahmanusta** [asciidoctor/asciidoctor#1796]

**Comment by rahmanusta** [comment]
Is it possible to access source_location and lineno in a BlockProcessor?

Thanks.

**Comment by robertpanzer** [comment]
Hi Rahman,

`Asciidoctor::AbstractBlock#source_location` gives you a cursor that allows you to access file and lineno.

See https://github.com/asciidoctor/asciidoctor/blob/master/lib/asciidoctor/reader.rb#L5 for the properties of a cursor and https://github.com/asciidoctor/asciidoctor/blob/master/lib/asciidoctor/abstract_block.rb#L26 for the declaration of the property on the block.

Cheers
Robert

**Comment by rahmanusta** [comment]
Hi @robertpanzer;

I already use this feature in a TreeProcessor, but when I tried I couldn't access lineno info in a blockmacro extension. Is it possible ?

Thanks.

**Comment by robertpanzer** [comment]
Oh yes, you’re right.
The source_location is set after the extensions have been called.
So you probably cannot access the source location in a block macro, sorry.

**Comment by rahmanusta** [comment]
So, is it possible to have this one?

Thanks.

**Comment by mojavelinux** [comment]
@robertpanzer is correct that this is not currently possible. It's a chicken egg problem.

The best solution I can see is to attach this information into the reader object that gets passed to the extension. The reader could have a parent, which would have the enclosing source range information.

**Comment by rahmanusta** [comment]
Thank you @mojavelinux, my main goal is about caching images generated for unnamed extension elements like stem:[x^2]. Now, I'm mapping binary images from md5 of formulas, but when we change formula, preview image is being not-found in AsciidocFX for short time till it is regenerated. However, If I have unique or close-to-unique key like lineno, I can increase UX more.

Theoretically, it is possible to get previous-block's lineno right ?

Thanks.

**Comment by mojavelinux** [comment]
Yes, you do have access to the parent, from which you can get the last child, which is the last child added to the tree.
